### PR TITLE
Fixes #31813 - unpin fog-core and net-ssh

### PR DESCRIPTION
**Pull request by ares**
CentOS Stream 8 already contains net-ssh 5.1.0 which conflicts with Foreman's
net-ssh 4.2.0. One can only perform dnf upgrade --nobest. The reason for
pinning old net-ssh was fog-core https://github.com/theforeman/foreman/commit/14e90094305a29440923a8fc21af1afde3173314

That seems to be fixed since fog-core 2.2.1 - https://github.com/fog/fog-core/commit/9640a4a40580191ddb92287f03f9b27864fc7f21

### CODE REVIEW

**Comment by ares** [comment]
I haven't done much testing however it's shortly after the branching, I think we should do this asap. I'm adding a packaging implications, however we don't need to bump version immediately. If we merge it, it would let developers install new versions and test. We can update packaging after a while

File: Gemfile
Comment (ekohl) [inline Gemfile]: Would it be good to specify an upper limit?

File: Gemfile
Comment (ezr-ondrej) [inline Gemfile:38]: Like this @ekohl? 6.1 is out, so we are going to test it with that, so we should just prevent installing new major version, right?
EDIT and it complies with [`net_scp`](https://rubygems.org/gems/net-scp)

**Review by ekohl — APPROVED** [review]

File: Gemfile
Comment (ekohl) [reply Gemfile:38]: Exactly.

**Comment by ezr-ondrej** [comment]
This: https://github.com/theforeman/foreman-packaging/pull/6293 should get in first, just to be sure we don't encounter the bug.

**Comment by ekohl** [comment]
> the bug.

Which bug?

**Comment by ezr-ondrej** [comment]
> Which bug?

The one referenced in description: https://github.com/theforeman/foreman/commit/14e90094305a29440923a8fc21af1afde3173314

**Comment by ezr-ondrej** [comment]
So if I understand correctly, we need a https://github.com/fog/fog-google/pull/422 to get in, so we can bump `fog-core`, so this needs to wait for it, otherwise we would break debian.

**Comment by ekohl** [comment]
Well, not really. https://github.com/fog/fog-google/blob/89acae970cea8307dc93d15be2f8b94aade089b6/fog-google.gemspec#L23-L24 should block a newer version getting in. That means it won't break, but you won't get a newer version either. That's why I'm fine with this change.

**Comment by ezr-ondrej** [comment]
> That means it won't break, but you won't get a newer version either.

I don't think so, I think that you won't get new version of fog-core, but you get a new version of `net-ssh` and that's an issue if I understand the description correctly.

**Comment by ares** [comment]
Re net-ssh upper limit. We don't specify it elsewhere and we didn't before the pin, which was supposed to be temporary. If we start using upper limits, we'll end up with old dependencies. I think it's better to find out in nightlies that some major version does not work than when we realize we need to upgrade `net-ssh` to newer version e.g. because of security vulnerability. We'd start to work in incompatibilities under time pressure. So I'd prefer to have no upper limit.

**Comment by ezr-ondrej** [comment]
> So I'd prefer to have no upper limit.

I'd agree, if we are upgrading the dependencies automatically in the packaging.
I'm not sure here, but I don't think it's happening, so having the upper limit is actually beter, because it makes the develop (+debian) env close to the EL production one. So we will be less likely to forget to update said dependencies.

I'd agree, that bast case would be to bump all dependencies in packaging once in a while (as part of a release process after branching maybe?) and then we would be probably better off to leave it up to the automated dependency resolution.

**Comment by ekohl** [comment]
I largely agree with @ezr-ondrej though I will note that an upper bound is good for stable branches. You really don't want to go back to stable branches and setting an upper bound when it does break.

And bumping post branching is a good practice and good to make it a (regular) process.

**Comment by ares** [comment]
Ok, I don't mind that much, IMHO we're changing the way it worked before the pin, without broader discussion of impacts, but I want to get this in ASAP after the branching.

FWIW while I agree with you both on ideal scenario and I wish there was a process to update all dependencies right after the branching, today reality is, there's no such process. Auto-updating of dependencies because no pins is something that works today already.

Anyway, is this pending packaging now? I think we don't need to update packaging to merge this necessarily.

**Comment by ekohl** [comment]
Given this still allows the prior versions, I'd say this isn't blocked on packaging which is why I approved it earlier. I was under the impression @ezr-ondrej was still testing new versions.

Order wise, IMHO it should be:
* Merge this
* Update requirements in foreman.spec (a script can do this correctly)
* Update gem RPM packages for the dependencies.

**Comment by ezr-ondrej** [comment]
I still belive this will break debian and dev. We will end up with fog-core 2.1 (due to fog-google) and net-ssh >5.1, which IIUC is incompatible, as commit mesage https://github.com/theforeman/foreman/commit/14e90094305a29440923a8fc21af1afde3173314 states (I didn't verify, but if the message is correct, we should not merge)

**Comment by ezr-ondrej** [comment]
This is still blocked by `fog-google` we need to resolve that prior merging this. I'll close for now.

**Pull request closed without merging**
